File: problem_028.c
// -*- compile-command: "gcc -o problem_028 problem_028.c -Wall" -*-
// Copyright (c) 2014 Michael Caldwell
#include <stdio.h>

int main()
{
    printf("Project Euler - Problem 28:\n"
           "Starting with the number 1 and moving to the right in a clockwise direction, form a 1001 by 1001 spiral. What is the sum of the numbers on the diagonals in this spiral?\n\n");

    int sum        = 0;
    int sideLength = 1001;
    int spiralSize = sideLength * sideLength;
    int position   = 0;
    int skipAmount = 1;
    int side       = 0;

    for (position=1 ; position<=spiralSize ; position++)
    {
        //printf("%d\n",position); // Debug
        sum += position;
        position += skipAmount;

        // Skip 2 positions every time we go around all 4 sides
        side++;
        if ( side % 4 == 0)
            skipAmount += 2;
    }

    printf("Sum: %d\n", sum);

    return 0;
}

File: problem_028.go
// -*- compile-command: "go build -o problem_028_go problem_028.go" -*-
// Copyright (c) 2018 Michael Caldwell
package main

import "fmt"

// 21 22 23 24 25
// 20 07 08 09 10
// 19 06 01 02 11
// 18 05 04 03 12
// 17 16 15 14 13

func main() {
	fmt.Println("Project Euler - Problem 28")
	fmt.Println("Starting with the number 1 and moving to the right in a clockwise direction, form a 1001 by 1001 spiral. What is the sum of the numbers on the diagonals in this spiral?\n")

	sum := 1
	pos := 1

	for i := 2; i <= 1001; i+=2 {
		for s := 0; s <4; s++ {
			pos += i
			sum += pos
		}
	}

	fmt.Println(sum)
}
